src/psyclone/tests/psyir/transformations/loop_fusion_test.py

```diff
@@ -0,0 +1,585 @@
+              do jj=1, n
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              do jj=1, n+1
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              end subroutine sub'''
+    with pytest.raises(TransformationError) as err:
+        fuse_loops(parser, code)
+    assert "Upper loop bounds must be identical, but are" in str(err.value)
+
+    # Test step size
+    code = '''subroutine sub()
+              integer :: ji, jj, n
+              integer, dimension(10,10) :: s, t
+              do jj=1, n, 2
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              do jj=1, n
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              end subroutine sub'''
+    with pytest.raises(TransformationError) as err:
+        fuse_loops(parser, code)
+    assert "Step size in loops must be identical, but are" in str(err.value)
+
+    # Test step size - make sure it defaults to 1
+    code = '''subroutine sub()
+              integer :: ji, jj, n
+              integer, dimension(10,10) :: s, t
+              do jj=1, n, 1
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              do jj=1, n, 1
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              end subroutine sub'''
+    fuse_loops(parser, code)
+
+
+# ----------------------------------------------------------------------------
+def test_fuse_different_loop_vars(parser):
+    '''
+    Test that loop variables are identicla
+    '''
+    code = '''subroutine sub()
+              integer :: ji, jj, n
+              integer, dimension(10,10) :: s, t
+              do jj=1, n
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              do ji=1, n
+                 do jj=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              end subroutine sub'''
+    with pytest.raises(TransformationError) as err:
+        fuse_loops(parser, code)
+    assert "Loop variables must be the same, but are 'jj' and 'ji'" \
+        in str(err.value)
+
+
+# ----------------------------------------------------------------------------
+@pytest.mark.xfail(reason="Needs evaluation of constant expressions")
+def test_fuse_correct_bounds(parser):
+    '''
+    Test that loop boundaries must be identical.
+    '''
+    # TODO: This test needs evaluation
+    # of constant expressions in PSyclone
+    code = '''subroutine sub()
+              integer :: ji, jj, n
+              integer, dimension(10,10) :: s, t
+              do jj=2-1, n
```

Comment:
The lower bounds of these two loops differ so the fusion should always fail, irrespective of evaluation of constant expressions. Perhaps the lower bounds should be the same?

Reply:
Yep